pkl-core/src/main/java/org/pkl/core/resource/ResourceReaders.java

```diff
@@ -488,7 +488,16 @@ public Optional<Object> read(URI uri)
       var dependency = getProjectDepsResolver().getResolvedDependency(assetUri.getPackageUri());
       var local = getLocalUri(dependency, assetUri);
       if (local != null) {
-        return VmContext.get(null).getResourceManager().read(local, null);
+        var resourceManager = VmContext.get(null).getResourceManager();
+        var securityManager = VmContext.get(null).getSecurityManager();
+        securityManager.checkReadResource(local);
+        var reader = resourceManager.getResourceReader(local);
+        if (reader == null) {
+          throw new VmExceptionBuilder()
+              .evalError("noResourceReaderRegistered", uri.getScheme())
+              .build();
+        }
+        return resourceManager.doRead(reader, local, null);
```

Comment:
The core problem is around using `ResourceManager#read`, which returns a `org.pkl.core.runtime.VmTyped` rather than a `org.pkl.core.resource.Resource`.

Additionally, we shouldn't use the ResourceManager here because that method statefully alters its cache. Since this method is already called within the same cache's `computeIfAbsent` call, mutating that cache results in a ConcurrentModificationException.